common/src/parse.cpp

```diff
@@ -79,60 +81,90 @@ pcl::console::parse_argument (int argc, const char * const * argv, const char *
 int
 pcl::console::parse_argument (int argc, const char * const * argv, const char * str, bool &val)
 {
+  long int dummy;
+  const auto ret = parse_argument (argc, argv, str, dummy);
+  if (ret != -1)
+  {
+    val = static_cast<bool> (dummy);
+  }
+  return ret;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+namespace pcl
+{
+namespace console
+{
+template <class T, class V = T(*)(const char*, const char**)> int
+parse_generic (V convert_func, int argc, const char* const* argv, const char* str, T& val)
+{
+  char *endptr = nullptr;
   int index = find_argument (argc, argv, str) + 1;
+  errno = 0;
 
   if (index > 0 && index < argc )
-    val = atoi (argv[index]) == 1;
+  {
+    val = convert_func (argv[index], &endptr);  // similar to strtol, strtod, strtof
+  }
+  // handle out-of-range, junk at the end and no conversion
+  if (errno == ERANGE || *endptr != '\0' || str == endptr)
+  {
+    return -1;
+  }
 
   return (index - 1);
 }
 
+template <class Integral> int
+parse_longint (int argc, const char * const * argv, const char * str, Integral &val)
+{
+  long int dummy;
+  const auto ret = parse_argument (argc, argv, str, dummy);
+  if ((ret == -1) ||  (dummy < std::numeric_limits<Integral>::min ()) || (dummy > std::numeric_limits<Integral>::max ()))
```

Comment:
Given the current CI failures on Windows I'm wondering if you need to explicitly cast
`static_cast<long int> (std::numeric_limits<Integral>::min ()))`

Reply:
Does this work on your setup? I don't have a windows compiler to check. If not, I'll try my luck with godbolt windows, but godbolt.ms seems very fragile to me. (2 diff issues filed in 2 days)

Reply:
It passes on my current setup (Ubuntu 18.04). By the way, this warning is spewing.
```
/home/sergio/Development/3rdparty/pcl/common/src/parse.cpp:99:1: warning: mangled name for ‘int pcl::console::parse_generic(V, int, const char* const*, const char*, T&) [with T = float; V = float (*)(const char*, char**) throw ()]’ will change in C++17 because the exception specification is part of a function type [-Wnoexcept-type]
 parse_generic (V convert_func, int argc, const char* const* argv, const char* str, T& val)
 ^~~~~~~~~~~~~
/home/sergio/Development/3rdparty/pcl/common/src/parse.cpp:99:1: warning: mangled name for ‘int pcl::console::parse_generic(V, int, const char* const*, const char*, T&) [with T = double; V = double (*)(const char*, char**) throw ()]’ will change in C++17 because the exception specification is part of a function type [-Wnoexcept-type]
```

Reply:
We can ignore the warning. Also, I was wondering if your setup was Windows. The tests pass on Ubuntu already 😄